homeassistant/components/history.py

```diff
@@ -119,19 +119,39 @@ def get_states(hass, utc_point_in_time, entity_ids=None, run=None,
     from sqlalchemy import and_, func
 
     with session_scope(hass=hass) as session:
-        most_recent_state_ids = session.query(
-            func.max(States.state_id).label('max_state_id')
-        ).filter(
-            (States.created >= run.start) &
-            (States.created < utc_point_in_time) &
-            (~States.domain.in_(IGNORE_DOMAINS)))
+        if entity_ids and len(entity_ids) == 1:
+            # Use an entirely different (and extremely fast) query if we only
+            # have a single entity id
+            most_recent_state_ids = session.query(
+                States.state_id.label('max_state_id')
+            ).filter(
```

Comment:
I just tested it in MySQL and I am seeing no performance impact to include it.

My preference would be to not do this unless we can think of an edge case where it will protect against something. I get the desire, but I worry we'd be sacrificing future readability of the code for a guard that may not be necessary...